### ISSUE TITLE
Expose ASGI scope on Request object

### ISSUE THREAD

**Issue by rmorshea** [sanic-org/sanic#2417]
**Is your feature request related to a problem? Please describe.**

I'm trying to write an application framework that works across ASGI compatible server implementations and to do this I'd like to make the ASGI scope available to users. Sanic does not make the scope publicly available though. 

**Describe the solution you'd like**
I'd like the `sanic.request.Request` object to provide a `scope` attribute which is the current `ASGIScope`

**Additional context**
Presently I'm working around this by accessing `request.app._asgi_app.transport.scope`

**Comment by ahopkins** [comment]
That's a good idea, and sounds easy enough that we can probably squeeze that in for the upcoming release.

**Comment by rmorshea** [comment]
You'll probably need to handle the case where the server isn't being run via ASGI, but from what I can tell it seems like the project is moving away from that.

**Comment by ahopkins** [comment]
> from what I can tell it seems like the project is moving away from running without ASGI.

Just curious, what makes you think this is the case? The built-in Sanic server (TMK) is still the most common deployment method. But, I do agree that ASGI support is an important alternative.

**Comment by rmorshea** [comment]
I can't recall exactly, but I may have skimmed [this thread](https://community.sanicframework.org/t/considering-asgi-support/70/3) and come away with the wrong idea.

**Comment by ahopkins** [comment]
No, not necessarily. ASGI is very important and I hope will continue to be. But we are not getting rid of the internal server either. Being geared specifically for Sanic it has the benefit of being optimized for performance in ways that ASGI servers cannot be.

**Comment by rmorshea** [comment]
That sounds reasonable, but then again, you could also make it a `@property` and return `None` if there's no `_asgi_app`.

I'm not very familiar with Sanic's internals though, so I can't say what the best option would be.

**Comment by azimovMichael** [comment]
Yeah I can do that, but I guess the downside is that currently the Request object is decoupled completely from any ASGI internals, and this change will change that a bit

**Comment by ahopkins** [comment]
@azimovMichael I would be happy for you to take this on. I see you have been assigned it already. Sorry for my delay, I am just now getting back to my regular schedule after some time off.

Instead of altering the `__init__`, we could just assign it:

```python
instance.request = request_class(
    ...
)
instance.request.stream = instance
...
instance.request._scope = scope
```

Using `@property` as suggested by @rmorshea is a nice idea so that it is less easily changed and a pattern used elsewhere on the `Request` object.

**Comment by ahopkins** [comment]
I would go so far as to do something like this:

```python
@property
def scope(self):
    if not self._scope:
        raise NotImplementedError("...")
    return self._scope
```

**Comment by ahopkins** [comment]
Actually... as I think about it more, I do not think that we even need to set it on the `Request` object at all. You should have access to it (I believe):

```python
request.transport.scope
```

Therefore, it is just a matter of returning that in the `property` and making sure that you are in ASGI.
